dlc-oracle/src/main/scala/org/bitcoins/dlc/oracle/DLCOracle.scala

```diff
@@ -0,0 +1,206 @@
+//    val coin = HDCoin(HDPurposes.SegWit, HDCoinType.fromNetwork(conf.network))
+//    val account = HDAccount(coin, 0)
+//    val chain = HDChain(HDChainType.External, account)
+//    HDAddress(chain, 0)
+//  }
+
+  // 585 is a random one I picked, unclaimed in https://github.com/satoshilabs/slips/blob/master/slip-0044.md
+  private val PURPOSE = 585
+
+  private val rValueAccount: HDAccount = {
+    val coin = HDCoin(HDPurpose(PURPOSE), HDCoinType.fromNetwork(conf.network))
+    HDAccount(coin, 0)
+  }
+
+  private val chainIndex = 0
+
+  private val signingKey: ECPrivateKey = {
+    val accountIndex = rValueAccount.index
+    val coin = rValueAccount.coin
+    val purpose = coin.purpose
+    // Use a different chain index so we don't reuse a key
+    val chain = chainIndex + 1
+
+    val path = BIP32Path.fromString(
+      s"m/${purpose.constant}'/${coin.coinType.toInt}'/$accountIndex'/$chain'/0'")
+
+    // what was previously used so we other wallets could find funds
+//    extPrivateKey
+//      .deriveChildPrivKey(SegWitHDPath(signingKeyHDAddress))
+//      .key
+
+    extPrivateKey.deriveChildPrivKey(path).key
+  }
+
+  val publicKey: SchnorrPublicKey = signingKey.schnorrPublicKey
+
+  def stakingAddress(network: BitcoinNetwork): Bech32Address =
+    Bech32Address(P2WPKHWitnessSPKV0(signingKey.publicKey), network)
+
+  protected[bitcoins] val rValueDAO: RValueDAO = RValueDAO()
+  protected[bitcoins] val eventDAO: EventDAO = EventDAO()
+  protected[bitcoins] val eventOutcomeDAO: EventOutcomeDAO = EventOutcomeDAO()
+
+  private def getPath(keyIndex: Int): BIP32Path = {
+    val accountIndex = rValueAccount.index
+    val coin = rValueAccount.coin
+    val purpose = coin.purpose
+
+    BIP32Path.fromString(
+      s"m/${purpose.constant}'/${coin.coinType.toInt}'/$accountIndex'/$chainIndex'/$keyIndex'")
+  }
+
+  private def getKValue(rValDb: RValueDb): ECPrivateKey =
+    getKValue(rValDb.label, rValDb.path)
+
+  private def getKValue(label: String, path: BIP32Path): ECPrivateKey = {
+    require(path.forall(_.hardened),
+            s"Cannot use a BIP32Path with unhardened nodes, got $path")
+    val priv = extPrivateKey.deriveChildPrivKey(path).key
+    val hash =
+      CryptoUtil.sha256(
+        priv.schnorrNonce.bytes ++ CryptoUtil.serializeForHash(label))
+    val tweak = ECPrivateKey(hash.bytes)
+
+    priv.add(tweak)
+  }
+
+  def listEventDbs(): Future[Vector[EventDb]] = eventDAO.findAll()
+
+  def listPendingEventDbs(): Future[Vector[EventDb]] = eventDAO.getPendingEvents
+
+  def listEvents(): Future[Vector[Event]] = {
+    for {
+      eventDbs <- eventDAO.findAll()
+      outcomes <- eventOutcomeDAO.findAll()
+    } yield {
+      val outcomesByNonce = outcomes.groupBy(_.nonce)
+      eventDbs.map(db => Event(db, outcomesByNonce(db.nonce)))
+    }
+  }
+
+  def createNewEvent(
+      label: String,
+      outcomes: Vector[String]): Future[EventDb] = {
+    for {
+      indexOpt <- rValueDAO.maxKeyIndex
```

Comment:
So I do worry about the usage of `maxKeyIndex`. There is nothing that protects against two separate threads calling `createNewEvent()` at the same time. If I understand correctly, this would lead to an r value being re-used? 

I think it would be wiser to have `rValueDAO` be responsible for retrieving _and_ incrementing the `maxKeyIndex`. We should then write test cases to make sure we can't get `rValueDAO` to return the same index on separte calls.

Reply:
I think it's not very vulnerable to that. The db has a primary key for the nonce so it should fail if you try to make a second with the same, also the r value is tweaked based on the label so it's extra safe there.